WIP: Document programmatic dependency resolution apis (#29930)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/dep-man/01-core-dependency-management/programmatic_dependency_resolution.adoc

====

Running this task, we get the following output:

----
include::{snippetsPath}/dependencyManagement/performingResolution-resolutionResult/tests/dot.out[]
----

Compare this to the output of the `dependencies` task:

----
include::{snippetsPath}/dependencyManagement/performingResolution-resolutionResult/tests/dependencies.out[]
----

Notice how the graph is the same for both representations.

[[sec:artifact_resolution]]
== Artifact resolution

After constructing a dependency graph, Gradle can perform artifact resolution on the resolved graph.
Gradle APIs can be used to influence the process of _artifact selection_ -- the mapping of a graph to a set of artifacts.
Gradle can then expose the results of artifact selection as an `ArtifactCollection`.
More commonly, however, the results are exposed as a `FileCollection`, which is a flat list of files that are downloaded in parallel.

Artifact selection operates on the dependency graph on a node-by-node basis.
Each node in the graph may expose multiple sets of artifacts, but only one of those sets may be selected.
For example, the `runtimeElements` variant of the Java plugins exposes a `jar`, `classes`, and `resources` artifact set.

> These three artifact sets represent the distributable, but in different forms.

Author: ```suggestion
These three artifact sets represent the same distributable, but in different forms.
```

For each node (variant) in a graph, Gradle performs attribute matching over each set of artifacts exposed by that node to determine the best artifact set.
If no artifact sets match the requested attributes, Gradle will attempt to construct an artifact transform chain to satisfy the request.

> For more details on the attribute matching process, see the <<variant_attributes.adoc#sec:attribute_matching,attribute matching>> section.
> 
> TODO: Link to attribute matching page for details.
> 
> === Implicit artifact selection

Author: ```suggestion
For more details on the attribute matching process, see the <<variant_attributes.adoc#sec:attribute_matching,attribute matching>> section.

=== Implicit artifact selection
```

.Artifact resolution task
====
include::sample[dir="snippets/dependencyManagement/performingResolution-artifactResolution/kotlin",files="build.gradle.kts[tags=artifact-resolution-task]"]
include::sample[dir="snippets/dependencyManagement/performingResolution-artifactResolution/groovy",files="build.gradle[tags=artifact-resolution-task]"]
====

This task is initialized similarly to the file resolution task:

.Wiring artifact resolution
====
include::sample[dir="snippets/dependencyManagement/performingResolution-artifactResolution/kotlin",files="build.gradle.kts[tags=implicit-artifact-resolution]"]
include::sample[dir="snippets/dependencyManagement/performingResolution-artifactResolution/groovy",files="build.gradle[tags=implicit-artifact-resolution]"]
====

Running this task, we can see that file metadata is included in the output:

----
include::{snippetsPath}/dependencyManagement/performingResolution-artifactResolution/tests/resolve-artifacts.out[]
----

=== Customizing artifact selection

In some cases, it is desirable to customize the selection process.
The `ArtifactView` API is the primary mechanism for influencing artifact selection in Gradle.

> `ArtifactView`s can:

Author: ```suggestion
An `ArtifactView` can:

```

- Trigger artifact transforms
- Select _alternative_ variants, such as sources or javadoc, for an entire resolution
- Perform _lenient_ artifact selection and resolution
- Filter selected artifacts

> NOTE: The `ArtifactView` can produce results as both `FileCollection`s and `ArtifactCollection`s.

Author: ```suggestion
NOTE: The `ArtifactView` can produce results as both a `FileCollection` and an `ArtifactCollection`.
```

> The below examples will only demonstrate using `FileCollection`s as the output.

Author: ```suggestion
The below examples will only demonstrate using a `FileCollection` as the output.
```

==== Triggering artifact transforms

> `ArtifactView`s can be used to trigger artifact selection using attributes different from those used to resolve the graph.

Author: ```suggestion
An `ArtifactView` can be used to trigger artifact selection using attributes different from those used to resolve the graph.
```

====

Using the `ArtifactView#withVariantReselection()` API, Gradle will optionally perform graph variant selection again before performing artifact selection on the new selected variant.
When Gradle selects artifacts for the `runtimeElements` node, it will use the attributes specified on the `ArtifactView` to reselect the graph variant, thus selecting the `sourcesElements` variant instead.
Then, traditional artifact selection will be performed on the `sourcesElements` variant to select the `jar` artifact set.

As a result, the sources jar is resolved for each node:

----
include::{snippetsPath}/dependencyManagement/performingResolution-artifactResolution/tests/resolve-sources.out[]
----

When this API is used, the attributes used for variant reselection are specified solely by the `ArtifactView#getAttributes()` method.
The graph resolution attributes specified on the configuration are completely ignored during variant reselection.

==== Performing lenient artifact resolution

The `ArtifactView` API can also be used perform lenient artifact resolution.

> This allows artifact resolution to be performed on a graph that contains failures -- for example when a request module was not found, the requested module version did not exist, or a conflict was not resolved.

Author: ```suggestion
This allows artifact resolution to be performed on a graph that contains failures -- for example when a requested module was not found, the requested module version did not exist, or a conflict was not resolved.
```
